reference/test/matrix/dense_kernels.cpp

```diff
@@ -2361,6 +2361,137 @@ TYPED_TEST(Dense, SquareMatrixGatherRowsIntoDenseFailsForWrongDimensions64)
+    gko::array<gko::int32> permute_idxs{exec, {2, 0}};
+
+    row_collection->row_scatter(&permute_idxs, this->mtx5);
+
+    GKO_ASSERT_MTX_NEAR(
+        this->mtx5, l<T>({{0.7, 1.1, 4.0}, {-2.0, 2.0, 4.5}, {3.0, 2.7, 6.5}}),
+        0.0);
+}
+
+
+TYPED_TEST(Dense, MatrixCanScatterRowsIntoDenseSubmatrix)
+{
+    using Mtx = typename TestFixture::Mtx;
+    using T = typename TestFixture::value_type;
+    auto exec = this->mtx5->get_executor();
+    auto row_collection = gko::initialize<Mtx>(I<I<T>>{{3.0, 2.7}}, exec);
+    gko::array<gko::int32> permute_idxs{exec, {0}};
+
+    row_collection->row_scatter(&permute_idxs,
+                                this->mtx5->create_submatrix({2}, {1, 3}));
+
+    GKO_ASSERT_MTX_NEAR(
+        this->mtx5,
+        l<T>({{1.0, -1.0, -0.5}, {-2.0, 2.0, 4.5}, {2.1, 3.0, 2.7}}), 0.0);
+}
+
+
+TYPED_TEST(Dense, MatrixScatterRowsFailsWithWrongDimensions)
+{
+    using Mtx = typename TestFixture::Mtx;
+    using T = typename TestFixture::value_type;
+    auto exec = this->mtx5->get_executor();
+    auto row_collection1 =
+        gko::initialize<Mtx>(I<I<T>>{{3.0, 2.7}, {0.7, 1.1}}, exec);
+    auto row_collection2 =
+        gko::initialize<Mtx>({{3.0, 2.7, 6.5}, {0.7, 1.1, 4.0}}, exec);
+    gko::array<gko::int32> permute_idxs1{exec, {2, 0}};
+    gko::array<gko::int32> permute_idxs2{exec, {1}};
+
+    ASSERT_THROW(row_collection1->row_scatter(&permute_idxs1, this->mtx5),
+                 gko::DimensionMismatch);
+    ASSERT_THROW(row_collection2->row_scatter(&permute_idxs2, this->mtx5),
+                 gko::DimensionMismatch);
+}
+
+
+TYPED_TEST(Dense, MatrixScatterRowsFailsWithInvalidState)
+{
+    using Mtx = typename TestFixture::Mtx;
+    using T = typename TestFixture::value_type;
+    auto exec = this->mtx5->get_executor();
+    auto row_collection =
+        gko::initialize<Mtx>({{3.0, 2.7, 6.5}, {0.7, 1.1, 4.0}}, exec);
+    gko::array<gko::int32> permute_idxs{exec, {200, 0}};
+
+    ASSERT_THROW(row_collection->row_scatter(&permute_idxs, this->mtx5),
+                 gko::InvalidStateError);
+}
+
+
+TYPED_TEST(Dense, MatrixCanScatterRowsUsingIndexSetIntoDense)
+{
+    using Mtx = typename TestFixture::Mtx;
+    using T = typename TestFixture::value_type;
+    auto exec = this->mtx5->get_executor();
+    auto mtx = gko::initialize<Mtx>({{2.2, 6.9, 7.8},
+                                     {4.7, 1.3, 7.6},
+                                     {9.2, 8.6, 4.5},
+                                     {8.1, 9.4, 6.8},
+                                     {9.6, 7.1, 2.5}},
+                                    exec);
+    auto row_collection = gko::initialize<Mtx>(
+        {{3.0, 2.7, 6.5}, {0.7, 1.1, 4.0}, {3.4, 3.8, 7.8}}, exec);
+    gko::index_set<gko::int32> permute_idxs{exec, {1, 0, 4}};
+
+    row_collection->row_scatter(&permute_idxs, mtx);
+
+    GKO_ASSERT_MTX_NEAR(mtx,
+                        l<T>({{3.0, 2.7, 6.5},
+                              {0.7, 1.1, 4.0},
```

Comment:
I am confused, looks to me like row 0 and 1 should be switched? Doesn't `permute_idxs{exec, {1, 0, 4}}` say `{3.0, 2.7, 6.5}` goes into row 1 ?

Reply:
An index set is always ordered. So `permute_idxs{exec, {1, 0, 4}}` is equivalent to `permute_idxs{exec, {0, 1, 4}}`. I will add a comment to `row_scatter`